// roots-wordpress/lib/index.js
// Generated by CoffeeScript 1.10.0
(function() {
  var API, RootsUtil, W, add_posts_to_locals, add_urls_to_posts, cloneDeep, find, includes, merge, path, render_single_views, request;

  API = require('./api');

  W = require('when');

  RootsUtil = require('roots-util');

  merge = require('lodash.merge');

  find = require('lodash.find');

  cloneDeep = require('lodash.clonedeep');

  includes = require('lodash.includes');

  path = require('path');

  module.exports = function(opts) {
    var RootsWordpress;
    if (opts == null) {
      opts = {};
    }
    if (!opts.site) {
      throw new Error('You must supply a site url or id');
    }
    if (!opts.post_types) {
      opts.post_types = {
        post: {}
      };
    }
    return RootsWordpress = (function() {
      function RootsWordpress(roots) {
        this.roots = roots;
        this.util = new RootsUtil(this.roots);
        opts.site = opts.site.replace(/http:\/\//, '');
      }

      RootsWordpress.prototype.setup = function() {
        var all, base, config, type;
        if ((base = this.roots.config).locals == null) {
          base.locals = {};
        }
        this.roots.config.locals.wordpress = {};
        all = (function() {
          var ref, results;
          ref = opts.post_types;
          results = [];
          for (type in ref) {
            config = ref[type];
            results.push(request(opts.site, type, config).then(render_single_views.bind(this, config, type)).then(add_urls_to_posts).then(add_posts_to_locals.bind(this, type)));
          }
          return results;
        }).call(this);
        return W.all(all);
      };

      return RootsWordpress;

    })();
  };

  request = function(site, type, config) {
    var params;
    params = merge({}, config, {
      type: type
    });
    delete params.template;
    return API({
      path: site + "/posts",
      params: params
    });
  };

  render_single_views = function(config, type, res) {
    var posts;
    posts = res.entity.posts;
    if (!config.template) {
      return {
        urls: [],
        posts: posts
      };
    }
    return W.map(posts, (function(_this) {
      return function(p) {
        var compiler, directory, extension, locals, outputFile, outputLink, tpl;
        directory = config.directory ? config.directory : type;
        tpl = path.join(_this.roots.root, config.template);
        locals = merge({}, _this.roots.config.locals, {
          post: p,
          wordpress:{[type]:posts}
        });
        //_this.roots.config.locals.wordpress[type] = posts;
        extension = typeof _this.roots.config.server === 'object' && _this.roots.config.server.clean_urls ? '' : '.html';
        outputFile = directory + "/" + p.slug + ".html";
        outputLink = directory + "/" + p.slug + extension;
        compiler = find(_this.roots.config.compilers, function(c) {
          return includes(c.extensions, path.extname(tpl).substring(1));
        });
        return compiler.renderFile(tpl, cloneDeep(locals)).then(function(res) {
          return _this.util.write(outputFile, res.result);
        })["yield"](outputLink);
      };
    })(this)).then(function(urls) {
      return {
        urls: urls,
        posts: posts
      };
    });
  };

  add_urls_to_posts = function(obj) {
    return obj.posts.map(function(post, i) {
      post._url = obj.urls[i];
      return post;
    });
  };

  add_posts_to_locals = function(type, posts) {
    return this.roots.config.locals.wordpress[type] = posts;
  };

}).call(this);
